Initial implementation of the new chunked upload

DeepDiver1975 opened the pull request:
as specified in https://dragotin.wordpress.com/2015/06/22/owncloud-chunking-ng/

@dragotin here we go

@PVince81 @icewind1991 @schiesbn @rullzer @blizzz @nickvergessen 

Please have a look - THX

Yes - more tests are to be written - on it

DeepDiver1975 committed:
Initial implementation of the new chunked upload - as specified in https://dragotin.wordpress.com/2015/06/22/owncloud-chunking-ng/

PVince81 commented:
How does this stream work and still exist between PHP requests ? I see fseek is disabled. Maybe not implemented yet ?

DeepDiver1975 commented:
> How does this stream work and still exist between PHP requests ? I see fseek is disabled. Maybe not implemented yet ?

the AssemblyStream will assembly the chunk files while the file is moved/copied to it's final destination.

There is no intent to have it available between requests.

Up to now I don't see the need to seek

PVince81 commented:
I still don't understand. Is the AssemblyStream used for a single chunk then, and it would save each chunk as separate files like we did in the past ?

So basically when uploading 3 chunks, the client does 3 PUT requests, each gets its own AssemblyStream that saves the chunk contents into a file.

But then who puts the chunks together ?

PVince81 commented:
How does this fit with:
`PUT remote.php/uploads/upload-id/chunk-id`
and
`MOVE remote.php/uploads/upload-id /path/to/target-file`

I can't seem to infer this from the code.

DeepDiver1975 commented:
the move is like this - https://github.com/DeepDiver1975/dav/blob/master/tests/client/upload.php#L33

DeepDiver1975 commented:
this magic .file file is implemented as this FutureFile which uses the AssemblyStream to stream all chunks to the destination.

PVince81 commented:
Okay, so is the AssemblyStream internally putting the chunks into that file at the right position, doing a fseek internally ? I suppose those "nodes" are actually the rnages ? (that's why I asked about fseek)

PVince81 commented:
So the AssemblyStream is actually simulating a single stream based on several chunk files, and is read by the MOVE operation (I thought AssemblyStream was for putting, not for reading, that's why I was confused). So the chunk files themselves simply get uploaded as if they were regular files.

Did I get it correctly ? :smile:

DeepDiver1975 commented:
> So the AssemblyStream is actually simulating a single stream based on several chunk files, and is read by the MOVE operation

yes

PVince81 commented on apps/dav/lib/upload/assemblystream.php:
```diff
@@ -0,0 +1,223 @@
+
+use Sabre\DAV\IFile;
+
+class AssemblyStream implements \Icewind\Streams\File {
```
Can you add a comment here to explain the AssemblyStream to avoid the many possible confusions ?

PVince81 commented on apps/dav/lib/upload/futurefile.php:
```diff
@@ -0,0 +1,81 @@
+use Sabre\DAV\Exception\Forbidden;
+use Sabre\DAV\IFile;
+
+class FutureFile implements \Sabre\DAV\IFile {
```
Add comment to explain what this is

PVince81 commented on apps/dav/lib/upload/uploadfolder.php:
```diff
@@ -0,0 +1,58 @@
+use OCA\DAV\Connector\Sabre\Directory;
+use Sabre\DAV\Exception\Forbidden;
+
+class UploadFolder implements \Sabre\DAV\ICollection {
```
and an explanation how the upload folder works. What elements does it contain ?

PVince81 commented:
Cool stuff :smile:

icewind1991 commented on apps/dav/lib/upload/assemblystream.php:
```diff
@@ -0,0 +1,223 @@
+class AssemblyStream implements \Icewind\Streams\File {
+
+	/** @var resource */
+	private $context;
```
I'm reasonably sure this needs to be public

DeepDiver1975 replied:
@icewind1991 can you elaborate? I did not face any issues having this private

karlitschek commented:
do we really want to do this for 9.0 ? Not sure we have the capacity

icewind1991 commented on apps/dav/lib/upload/assemblystream.php:
```diff
@@ -0,0 +1,223 @@
+
+		// sort the nodes
+		$nodes = $this->nodes;
+		@usort($nodes, function(IFile $a, IFile $b) {
```
Why the `@`?

DeepDiver1975 replied:
No idea ;-)

DeepDiver1975 replied:
ah - because of that:

usort(): Array was modified by the user comparison function

DeepDiver1975 commented:
@karlitschek the code is ready since month. No reasons to wait any longer from my pov.

schiesbn commented:
did you already tried this with encryption enabled? Wonder how this works if uploaded chunks are not meet the expected chunk size of a encryption block? For the last chunk of a file encryption apply padding to meet the expected size. So maybe we will just end up with some additional padded chunks in the middle of the file. So it could work, just increase the size of the encrypted file a bit more... but will fseek work on such a file? We need to test this in detail if not already happened.

DeepDiver1975 commented:
> did you already tried this with encryption enabled? Wonder how this works if uploaded chunks are not meet the expected chunk size of a encryption block? For the last chunk of a file encryption apply padding to meet the expected size. So maybe we will just end up with some additional padded chunks in the middle of the file. So it could work, just increase the size of the encrypted file a bit... but will fseek work on such a file? We need to test this in detail if not already happened.

Nice input - thanks a lot! We need to add tests with encryption enabled.

blizzz commented on apps/dav/lib/upload/assemblystream.php:
```diff
@@ -0,0 +1,223 @@
+
+		list($node, $posInNode) = $this->getNodeForPosition($this->pos);
+		if (is_null($node)) {
+			return null;
```
not a string as indicated in PHP Doc

blizzz commented on apps/dav/lib/upload/assemblystream.php:
```diff
@@ -0,0 +1,223 @@
+	 * @return bool
+	 */
+	public function stream_close() {
+	}
```
PHP Doc wants to return bool not null

blizzz commented on apps/dav/lib/upload/uploadfolder.php:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	function createFile($name, $data = null) {
+		// TODO: verify name - should be a simple number
```
intend to complete this before merging?

blizzz commented on apps/dav/lib/upload/uploadfolder.php:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	function childExists($name) {
+		if ($name === '.file') {
```
maybe we wanna put that '.file' also into a constant somewhere?

blizzz commented:
Overall this looks nice and clean, though information is a bit sparse, as pointed out by @PVince81.

DeepDiver1975 commented:
@schiesbn this work with encryption enabled - current drawback: the chunks are not encrypted

blizzz commented:
As long as the upload files stay on the local server i guess it is OK, due to the threat model. Isn't it?

DeepDiver1975 commented:
> As long as the upload files stay on the local server i guess it is OK, due to the threat model. Isn't it?

No idea ... @LukasReschke @schiesbn THX

DeepDiver1975 commented:
Should we name the resource upload or uploads?

@dragotin @icewind1991 @blizzz @PVince81 @schiesbn

DeepDiver1975 commented:
> though information is a bit sparse, as pointed out by @PVince81.

@blizzz @PVince81 some more information has been added - let me know is you'd like to have more?

DeepDiver1975 commented:
@PVince81 as soon as this is merged we can start with a chunked upload in the files app - would be cool to see that working

PVince81 commented:
>Should we name the resource upload or uploads?

It's more of a Webdav convention. Are collections plural ? @evert

LukasReschke commented:
> > As long as the upload files stay on the local server i guess it is OK, due to the threat model. Isn't it?
> 
> No idea ... @LukasReschke @schiesbn THX

From my PoV acceptable and we can document it as-is but I see already potentially people complaining about it. Especially if they have some custom encryption modules implemented. But I can't judge how much effort this would be to implement in the future.

PVince81 commented on apps/dav/lib/upload/assemblystream.php:
```diff
@@ -0,0 +1,231 @@
+/**
+ * Class AssemblyStream
+ *
+ * The assembly stream is used to assembly file chunks while the file is read.
```
=> "The assembly stream is a virtual strip that wraps multiple chunks. Reading from the stream transparently accessed the underlying chunks and give a representation as if they were already merged together"

DeepDiver1975 replied:
THX - I'm good at coding but not at documenting - much appreachiated

DeepDiver1975 replied:
> I'm good at coding

okay - selfish self-observation :speak_no_evil:

blizzz replied:
not suffering imposter syndrome, eh ;)

PVince81 commented on apps/dav/lib/upload/futurefile.php:
```diff
@@ -0,0 +1,97 @@
+ *
+ * The FutureFile is a SabreDav IFile which connects the chunked upload directory
+ * with the AssemblyStream, who does the final assembly job
+ *
```
Ok so basically this is the virtual file that is composed of the chunks

DeepDiver1975 replied:
yes

schiesbn commented:
>> As long as the upload files stay on the local server i guess it is OK, due to the threat model. Isn't it?

> No idea ... @LukasReschke @schiesbn THX

I only see a problem if a different primary storage is used. This way the unencrypted chunks would be stored on a external server for some time which is probably not acceptable. In all other use cases it might be OK

DeepDiver1975 commented:
> From my PoV acceptable and we can document it as-is but I see already potentially people complaining about it. Especially if they have some custom encryption modules implemented. But I can't judge how much effort this would be to implement in the future.

Chunks of the "old" chunked upload are not encrypted either -> no issue

DeepDiver1975 commented:
> I only see a problem if a different primary storage is used.

objectstore and encryption are anyhow not supported at the moment :see_no_evil:

karlitschek commented:
@DeepDiver1975 Well it's not so easy. :-) This needs to be implemented on the client too, needs a ton of testing and even the current code needs some discussion as seen here in the PR. So can we please talk about this before we push this into the 9.0 and 2.1 timeframe and risk other 9.0 features?
@cmonteroluque

cmonteroluque commented:
@DeepDiver1975 and all: @karlitschek and I just spoke about it. This is not part of the 9.0 plan, which is likely to be fully loaded already (especially as we look into bugs that must be fixed in conjunction with the architectural and functionality changes already in 9.0). I am moving this into the backlog and we'll revise it either if all 9.0 functionality is already delivered, or as part of 9.1. @MTRichards fyi

cmonteroluque commented:
added to the internal tracking list for 9.1 (unless we can bring it in earlier somehow)

karlitschek commented:
Yes. Definitely something to explore in the future. Not sure if we have time in 9.0

evert commented:
>> Should we name the resource upload or uploads?

> It's more of a Webdav convention. Are collections plural ? @evert

I have no opinion =)

DeepDiver1975 commented:
@karlitschek reopened - based on our talk yesterday

karlitschek commented:
Yes. Thanks. Still a good thing just a bit lower priority then the rest of the 9 stuff

dragotin commented:
I realised that new WebDAV endpoints are part of the 9.0 feature plan, but how does that help (the client at least) if the new endpoint is not complete? We can only use it if it implements everything, also the new chunking (assuming that we agreed on it, I never got real feedback from @karlitschek )

karlitschek commented:
We have to discuss this new chunked uploading and decide if we want to do this or not. Sorry but this never happened. The decision to move to a new WebDAV endpoint to 9 or later is independent.

DeepDiver1975 commented:
Can we put this to the agenda for next week? @karlitschek - THX

karlitschek commented:
yes. have to!

DeepDiver1975 commented:
> yes. have to!

THX

DeepDiver1975 commented:
rebased - just to not let the diff get to big .... still on backlog with potential merge into 9.0 if time permits

butonic commented:
For objectstore there are PRs from People who know more about ceph than me that split bigger files into 4MB objects. Swift even has special large object support: http://docs.openstack.org/developer/swift/overview_large_objects.html

@DeepDiver1975 jost to Keep in mind that assembly of chunked uploads should be passed down to the storage API so that we can directly upload the chunks as large file Segments, without having to assemble ad disassemble large uploads in temp space.

@dragotin Also 4MB chunks (instead of 10MB) are a nice size that we can keep in memory (default php temp streams are 2MB) which is getting more important with concurrent uploads of bigger files.

icewind1991 commented:
> Also 4MB chunks (instead of 10MB) are a nice size

Note that smaller chunks (meaning more chunks) will have a noticeable negative effect on performance due to overhead for every chunk

guruz commented:
@butonic @icewind1991 FYI We're already using 5 MB chunks in the client since some time. We're also doing parallel upload of 3 chunks at once for recent server versions. If the server performance keeps improving, we're thinking of sending more chunks concurrently (depending on bandwidth estimates)

DeepDiver1975 commented:
> @DeepDiver1975 jost to Keep in mind that assembly of chunked uploads should be passed down to the storage API so that we can directly upload the chunks as large file Segments, without having to assemble ad disassemble large uploads in temp space.

We had many discussion and thoughts about the chunked upload process - we explicitly decided against a storage based chunked upload because of transactional/locking reasons.

Furthermore for some operations on the server we have to have the full file at hand before pushing it to the final storage.

butonic commented:
Yes, encryption comes to mind.

DeepDiver1975 commented:
the final move/copy operation requires the same execution time - independent of the amount of chunks:

````
deepdiver@alien:~/Development/ownCloud/master-autotest/apps/dav/tests/perf [chunked-upload-dav]$ php ./chunkperf.php http://localhost:8888/remote.php/dav admin admin test.jpg 2
MKCOL http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c2feddfda9.87703547 ... 201 - 0.70805406570435 seconds
PUT http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c2feddfda9.87703547/0 ... 201 - 1.6887340545654 seconds
PUT http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c2feddfda9.87703547/1 ... 201 - 1.7014858722687 seconds
PUT http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c2feddfda9.87703547/2 ... 201 - 1.6527581214905 seconds
PUT http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c2feddfda9.87703547/3 ... 201 - 1.713751077652 seconds
PUT http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c2feddfda9.87703547/4 ... 201 - 1.7588200569153 seconds
PUT http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c2feddfda9.87703547/5 ... 201 - 1.6662499904633 seconds
PUT http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c2feddfda9.87703547/6 ... 201 - 1.6596500873566 seconds
PUT http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c2feddfda9.87703547/7 ... 201 - 1.6562550067902 seconds
PUT http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c2feddfda9.87703547/8 ... 201 - 1.7901430130005 seconds
COPY http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c2feddfda9.87703547/.file ... 204 - 9.548593044281 seconds

deepdiver@alien:~/Development/ownCloud/master-autotest/apps/dav/tests/perf [chunked-upload-dav]$ php ./chunkperf.php http://localhost:8888/remote.php/dav admin admin test.jpg 20
MKCOL http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c334c11948.84484702 ... 201 - 0.74006199836731 seconds
PUT http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c334c11948.84484702/0 ... 201 - 1.8090200424194 seconds
COPY http://localhost:8888/remote.php/dav/uploads/admin/transfer5698c334c11948.84484702/.file ... 204 - 9.6941208839417 seconds

DeepDiver1975 commented:
It time to move this in - please approve - THX

@PVince81 @LukasReschke @schiesbn @icewind1991 @MorrisJobke @dragotin THX

rullzer commented:
Looks good.

@DeepDiver1975 Shall I adopt my recently added intergration tests to this new chunking?

DeepDiver1975 commented:
> Shall I adopt my recently added intergration tests to this new chunking?

Would be awesome! Maybe the c++ perf tests as well? THX a lot!

icewind1991 commented on apps/dav/lib/upload/futurefile.php:
```diff
@@ -0,0 +1,103 @@
+	 */
+	function getSize() {
+		$children = $this->root->getChildren();
+		$sizes = array_map(function($node) {
```
```php
return array_reduce($children, function($size + $node) {return $size + $node->getSize}, 0);
```

DeepDiver1975 replied:
> $size + $node

:question:

icewind1991 replied:
`$size + $node->getSize()`

DeepDiver1975 replied:
:see_no_evil: feel free to commit and push - THX

rullzer committed:
Add intergration tests for new chunking

The same intergration tests as for the old endpoint.
But now using the new chunking.

We upload 3 chunks in different order and the result should be the same
in all three.

rullzer commented:
@DeepDiver1975 the last step (the move) gives me a 404 here. (also happens in your perf example)

DeepDiver1975 commented:
> @DeepDiver1975 the last step (the move) gives me a 404 here. (also happens in your perf example)

grrrr - not again - I was analysing and fixing this already a couple of times ..... :angry:

DeepDiver1975 commented:
@rullzer looking forward to your integration tests - THX

rullzer commented:
It fails at https://github.com/owncloud/core/blob/master/apps/dav/lib/connector/sabre/filesplugin.php#L147-L149 which are checks writen for the old endpoint... so that might explain it...

rullzer commented:
Ok intergration tests added. If I comment out the checkMove test they work. So it is a matter of fixing that now ;)

rullzer commented:
@DeepDiver1975 https://github.com/owncloud/core/commit/e987d8cd47b5c8ea47b8b77b056343c32cf23666 that broke it...

MTRichards commented:
Estimation: 2 Days

DeepDiver1975 committed:
Fix chunk file move

DeepDiver1975 committed:
Fix unit test of file plugin

DeepDiver1975 commented:
Unit and integration tests are passing - let's move this in

DeepDiver1975 commented:
@rullzer @dragotin final review please - thx

dragotin commented:
For the record: We do not have the implementation on the client yet. It is not likely that it will be finished before this is released with the server.

DeepDiver1975 commented:
> For the record: We do not have the implementation on the client yet. It is not likely that it will be finished before this is released with the server.

as soon as there is anything in the client available we will start testing asap - got that.
In addition we will enhance our own integration tests over the upcoming months

dragotin commented:
integration test cool

rullzer commented on apps/dav/lib/connector/sabre/filesplugin.php:
```diff
@@ -146,11 +148,17 @@ public function initialize(\Sabre\DAV\Server $server) {
 	function checkMove($source, $destination) {
+		$sourceNode = $this->server->tree->getNodeForPath($source);
+		if ($sourceNode instanceof FutureFile) {
+			return;
```
Isn't that a bit to liberal? What if the file we are uploading to does not allow updates?

DeepDiver1975 replied:
This is a valid scenario we need to cover in the second part of the implementation which is know as 'Announcing the upload'

rullzer replied:
Ah OKIDOKI then.

rullzer commented:
Lets do this! :+1:

DeepDiver1975 commented:
THX

rperezb commented:
@DeepDiver1975 @dragotin from a qa pov, is it needed further testing in addition to integration ones?
thx
@owncloud/qa

DeepDiver1975 commented:
> from a qa pov, is it needed further testing in addition to integration ones?
> thx

no - we need a client supporting this to setup all the smashboxing and so on

rperezb commented:
@mcastroSG this is of your interest, perhaps you may help with the smashbox too

DeepDiver1975 commented:
> perhaps you may help with the smashbox too

hold your horses - as said we first need a client supporting the new chunking and use the new webdav endpoint

SergioBertolinSG commented:
Apparently ownbrander can be used for creating a client with a custom endpoint. Not sure if that can be used with smashbox in jenkins though.

DeepDiver1975 commented:
Ownbrander will not help. Support needs to be added to the client.
